Allow a saga to be started by multiple message types

arootbeer opened the pull request:
Implementation for #100

arootbeer committed:
Allow a saga to be started by multiple message types

mizrael commented on tests/OpenSleigh.Tests/SagaDescriptorTests.cs:
can you test what happens when a saga has more than one initiator

mizrael commented:
can you test the case of a saga receiving all the initiator messages sequentially/in parallel?

arootbeer commented:
> can you test the case of a saga receiving all the initiator messages sequentially/in parallel?

That's changing this how?

https://github.com/mizrael/OpenSleigh/blob/9652aad6db16a600a91219ca4071c0950608ec73/tests/OpenSleigh.Tests/SagaInstanceTests.cs#L242-L274

mizrael commented:
> > can you test the case of a saga receiving all the initiator messages sequentially/in parallel?
> 
> That's changing this how?
> 
> https://github.com/mizrael/OpenSleigh/blob/9652aad6db16a600a91219ca4071c0950608ec73/tests/OpenSleigh.Tests/SagaInstanceTests.cs#L242-L274

we'd need E2E tests as well

arootbeer committed:
E2E tests for in-memory transport and persistence

arootbeer commented:
@mizrael please review the test I added

I don't have infrastructure locally to test SQL, Kafka, or RabbitMQ, so I'd appreciate it if you'd wire those up and test them. I also left a set of commented assertions here for your input

https://github.com/mizrael/OpenSleigh/blob/335e033e7cdc5b7f59e03ca5e8c5d65ca8ca53d1/tests/OpenSleigh.E2ETests/SimpleMultiStartSagaScenario.cs#L83-L87

mizrael commented:
> @mizrael please review the test I added
> 
> I don't have infrastructure locally to test SQL, Kafka, or RabbitMQ, so I'd appreciate it if you'd wire those up and test them. I also left a set of commented assertions here for your input
> 
> https://github.com/mizrael/OpenSleigh/blob/335e033e7cdc5b7f59e03ca5e8c5d65ca8ca53d1/tests/OpenSleigh.E2ETests/SimpleMultiStartSagaScenario.cs#L83-L87

thanks! I am about to catch a flight, it might take a bit before I can finalize the review.

Just FYI, if you have Docker installed, you can use this https://github.com/mizrael/OpenSleigh/blob/develop/tests/infra/docker-compose.yml to spin up all the infra necessary to run the tests

arootbeer commented:
Not sure if something's changed since you built the E2E container stack, but `bitnami/kafka` [no longer has any tags on DockerHub](https://hub.docker.com/r/bitnami/kafka/tags/)

<img width="1412" height="161" alt="image" src="https://github.com/user-attachments/assets/c8144441-5c48-4a2c-9d51-8a89d4cd2968" />

arootbeer commented:
I've asked @iseesharp83 to pick up any further work on this PR in the near term

iseesharp83 commented:
@mizrael I've been trying to get the e2e tests to run, and while replacing `bitnami` with `bitnamilegacy` worked for the containers, I'm getting errors like `Cannot consume scoped service 'OpenSleigh.Transport.Kafka.IKafkaMessageHandler' from singleton 'OpenSleigh.Transport.IMessageSubscriber`1[OpenSleigh.E2ETests.StartSimpleSaga]'.` - this is in the SqlKafkaSimpleSagaScenario. The registration of IKafkaMessageHandler and IMessageSubscriber<> happen in the `UseKafkaTransport` extension. For now I'll skip these though and verify with an E2E test that we can add multiple messages, out-of-order, and have them start all the same.

iseesharp83 commented:
Actually I verified that @arootbeer added the e2e tests in the InMemory section and they're passing. I've pushed a change to use the `bitnamilegacy` so it's able to run.

mizrael commented:
hi @iseesharp83 , thanks for your patience! I've updated the docker compose definition and made some fixes to the existing E2E tests. Can you please merge down from `develop` and try again?

arootbeer commented:
@mizrael thanks for the docker compose updates! I've integrated them locally, and everything appears to start up correctly now. 

I had to update the E2E tests to use the PostgreSQL persistence layer in order to get the SqlRabbit fixture to run. The SqlKafka fixture _does not_ run, giving this exception:

```System.InvalidOperationException: Cannot consume scoped service 'OpenSleigh.Transport.Kafka.IKafkaMessageHandler' from singleton 'OpenSleigh.Transport...

System.InvalidOperationException
Cannot consume scoped service 'OpenSleigh.Transport.Kafka.IKafkaMessageHandler' from singleton 'OpenSleigh.Transport.IMessageSubscriber`1[OpenSleigh.E2ETests.StartSimpleSaga]'.
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.VisitCallSite(ServiceCallSite callSite, CallSiteValidatorState argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.VisitConstructor(ConstructorCallSite constructorCallSite, CallSiteValidatorState state)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.VisitRootCache(ServiceCallSite singletonCallSite, CallSiteValidatorState state)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.VisitCallSite(ServiceCallSite callSite, CallSiteValidatorState argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.ValidateCallSite(ServiceCallSite callSite)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.OnCreate(ServiceCallSite callSite)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(ServiceIdentifier serviceIdentifier)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at OpenSleigh.SubscribersBackgroundService.<>c__DisplayClass6_0.<ExecuteAsync>b__0(Type messageType) in C:\codez\OpenSleigh\src\OpenSleigh\SubscribersBackgroundService.cs:line 43
   at System.Linq.Enumerable.SelectIListIterator`2.Fill(IList`1 source, Span`1 results, Func`2 func)
   at System.Linq.Enumerable.SelectIListIterator`2.ToArray()
   at OpenSleigh.SubscribersBackgroundService.ExecuteAsync(CancellationToken stoppingToken) in C:\codez\OpenSleigh\src\OpenSleigh\SubscribersBackgroundService.cs:line 40
   at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>b__14_1(IHostedService service, CancellationToken token)
   at Microsoft.Extensions.Hosting.Internal.Host.ForeachService[T](IEnumerable`1 services, CancellationToken token, Boolean concurrent, Boolean abortOnFirstException, List`1 exceptions, Func`3 operation)
   at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>g__LogAndRethrow|14_3(<>c__DisplayClass14_0&)
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at OpenSleigh.E2ETests.E2ETestsBase.<>c__DisplayClass1_0.<<RunScenarioAsync>b__0>d.MoveNext() in C:\codez\OpenSleigh\tests\OpenSleigh.E2ETests\E2ETestsBase.cs:line 39
--- End of stack trace from previous location ---
   at OpenSleigh.E2ETests.E2ETestsBase.RunScenarioAsync(Int32 hostsCount, Action`2 configureServices, Func`2 runner, CancellationTokenSource runningTokenSource) in C:\codez\OpenSleigh\tests\OpenSleigh.E2ETests\E2ETestsBase.cs:line 43
   at OpenSleigh.E2ETests.SimpleSagaScenario.run_single_message_scenario(Int32 hostsCount) in C:\codez\OpenSleigh\tests\OpenSleigh.E2ETests\SimpleSagaScenario.cs:line 33
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass47_0.<<InvokeTestMethodAsync>b__1>d.MoveNext() in /_/src/xunit.execution/Sdk/Frameworks/Runners/TestInvoker.cs:line 259
--- End of stack trace from previous location ---
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in /_/src/xunit.execution/Sdk/Frameworks/ExecutionTimer.cs:line 48
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in /_/src/xunit.core/Sdk/ExceptionAggregator.cs:line 90
```

Which is easily enough explained by the code:

https://github.com/mizrael/OpenSleigh/blob/7cfded408461289e56a892669f66b086263cfa1b/src/OpenSleigh.Transport.Kafka/IBusConfiguratorExtensions.cs#L54

https://github.com/mizrael/OpenSleigh/blob/7cfded408461289e56a892669f66b086263cfa1b/src/OpenSleigh.Transport.Kafka/IBusConfiguratorExtensions.cs#L58

Since this PR is about multiple messages, I'm going to work on expanding the SqlRabbit test suite to run the simple multi-start saga scenario next. If you've got bandwidth to figure out what the above registrations are supposed to look like, I'm happy to do the same for SqlKafka.

mizrael commented:
that's weird. The Kafka E2E tests run just fine on my setup
<img width="333" height="125" alt="image" src="https://github.com/user-attachments/assets/f76a5f13-21eb-4d2e-a31e-b15d4d6c5266" />


can you try deleting all the volumes and containers you have and recreate the infra?

mizrael commented:
on a side note, resolving a transient service from a singleton one is definitely a code smell. I'll see what I can do on that one. Thanks for pointing this out!

arootbeer commented:
> that's weird. The Kafka E2E tests run just fine on my setup
> 
> can you try deleting all the volumes and containers you have and recreate the infra?

It shouldn't have anything whatsoever to do with the Kafka infrastructure, because it's failing to instantiate the `SubscribersBackgroundService` instance needed to actually start the host

<img width="808" height="378" alt="image" src="https://github.com/user-attachments/assets/b1aa6812-4923-4fd0-8c15-93d3d942f036" />

Which...makes it all the more odd that it's working fine for you, because I haven't made any changes whatsoever to anything Kafka-related in this PR, nor have I updated any dependency versions in any of the projects.

mizrael commented:
That's weird indeed. Anyways, I started a semi-big refactoring of the message processors, I should be able to merge the changes in a day or two. It shouldn't really impact your changes though

arootbeer committed:
E2E tests for optimistic concurrency

arootbeer commented:
> Since this PR is about multiple messages, I'm going to work on expanding the SqlRabbit test suite to run the simple multi-start saga scenario next. 

This worked fine - as long as the messages are published with enough time in between that we don't have to worry about optimistic concurrency issues.

If I publish two messages at the same time, they are processed in parallel. One of them will start the saga, but the second one fails and is not getting retried successfully even though it appears that it is being re-enqueued correctly. I have added signaling for optimistic concurrency in `SqlSagaStateRepository.LockAsyncCore(...)` and `InMemorySagaStateRepository.LockAsync(...)` via a new `OptimisticLockException` class, and I've added handling for that in `RabbitMessageSubscriber` and `InMemorySubscriber`. 

I didn't see existing handling for `LockException` in `KafkaMessageSubscriber`, so I didn't attempt to add support there.

mizrael committed:
deleted file
